cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/AppMetadataStore.java

```diff
@@ -655,7 +635,8 @@ public void updateAppSpec(ApplicationId appId, ApplicationSpecification spec) th
     if (LOG.isTraceEnabled()) {
       LOG.trace("App spec to be updated: id: {}: spec: {}", appId, GSON.toJson(spec));
     }
-    ApplicationMeta existing = getApplication(appId);
+    ApplicationMeta existing = ApplicationId.DEFAULT_VERSION.equals(appId.getVersion())
```

Comment:
this shouldn't happen here. The only place we should be mapping snapshot to latest is at the very start, in the handler. In those cases, ApplicationReference and ProgramReference should be used. Wherever ApplicationId is used, it should be an actual id.

Reply:
similar comments everywhere else. I'm not really understanding why this mapping is happening all over the code, is there something that is forcing us to do this?

Reply:
I updated it in the callers of getApplication. OK I'll back traverse more and change it from handler side. And it needs more careful examination I guess.